### DISCUSSION TITLE
How to automatically remember page position and other states per file?

### DISCUSSION THREAD

**Discussion by cesar-scapella** [sumatrapdfreader/sumatrapdf#3994] [Documentation]
Hi guys, it is a super handy function that I saw in other PDF viewers but can't find it in Sumatra.

Is there a setting to make Sumatra automatically store the page we are in when we close?

I am aware of bookmarks (that is not it).
It is a function that once you set, you don't need to do anything more, just close the file normally and as you open it another day, it automatically goes to the last page you were in, without needing any command.

I am also aware of the setting "NoPage" in the settings file but seems like you need to go there and type it (it is not automatic), and that is kind of unpractical when someone is dealing with various files per day (I browse several different books for studying).

I know Foxit has this function but I would love to not go there (heavy bloated proprietary software).
I love Sumatra: clean, efficient, light...

That is it guys, I hope you can help
Thanks

**Comment by GitHubRulesOK** [comment]
oddly it is called remember ?
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/130f9ed6-013a-4cbf-895e-6f772b95e205)

**Reply by cesar-scapella** [reply]
Well, I tinkered with these options a lot, they don't work for remembering last page...
Thanks.

**Comment by Peter-202122** [comment]
For me the odd thing (concerning the terms in the window shown by GitHubRulesOK) is that the option is called _"Remember **these** settings for each document"_.

The word "these" obviously refers (only) to the settings mentioned above this option, which are "Layout, "Zoom" and presumably the showing of the bookmarks sidebar.

So checking this box one would not assume that it also covers cesar-scapella's request (namely remembering the last opened page within a document). 

Indeed it works when checking the box. (I also prefer that the last position is remembered, so I can tell from my own experience.) But from a logical point of view (and for easier finding this function), wouldn't it be necessary to have an own checkbox concerning explicitly the remembering of the last opened page?

PS:
In the German translation the option is described in a better way:
Instead of _"Remember **these** settings for each document"_ the German version says:
_"Ansicht-Einstellungen für jedes Dokument einzeln merken"_
which would mean in English:
_"Remember View-Settings for each document separately"_

So the word "these" is avoided in the German version. And with some fantasy, as "View-Settings" one could also regard the last page opened. So that's better than the English version. But nevertheless I think that an own checkbox would be the clearest solution.

**Reply by cesar-scapella** [reply]
Hi, I already tinkered with these options, unfortunately it doesn't do what I mention.
It is so odd if it hasn't already been implemented... such a handy-convenient function.

**Reply by GitHubRulesOK** [reply]
with the 2 settings highlighted (close SumatraPDF and restart to trigger them) the file states are remembered in settings file which includes last page viewed
```
FileStates [
	[
		FilePath = C:\Users\lez\Downloads\1-27.pdf
		Favorites [
		]
		IsPinned = false
		IsMissing = false
		OpenCount = 2
		UseDefaultState = false
		DisplayMode = continuous
		ScrollPos = -1 -1
		PageNo = 225
```

similar data is held in session states
```
SessionData [
	[
		TabStates [
			[
				FilePath = C:\Users\lez\Downloads\1-27.pdf
				DisplayMode = continuous
				PageNo = 225
```

**Reply by cesar-scapella** [reply]
Yes, it works. I was avoiding using theses settings because it bring another issues... But good news, after tinkering a bit more, I think I figured what is happening. I will explain in another reply (to my initial post).

**Comment by cesar-scapella** [comment]
Thanks to @GitHubRulesOK and @Peter-202122 for replying...

I think I figured what is happening.

As pointed out by @GitHubRulesOK, when we enable both settings below, it works.
The settings are:
"Remember these settings for each document"
"Remember opened files"

However, it brings another issue for those using Sumatra with tabs turned off. So, whenever I opened one file it also opens a second windows and re-opens the last file previously opened in Sumatra, and of course that is not what the user wants. Finally today I understood why such a weird behavior: it is simple because I was always using ESC to close any number of opened files in Sumatra. It appears as though ESC works well for closing any file except the last one (in case we have several Sumatra windows opened). Of course it closes the Sumatra application but it seems as though this last file is considered "opened" and Sumatra will automatically bring it back as soon as the user open some other file on another day.

So I noticed that if I close that last file by using CTRL + W (or by File -> Close) then Sumatra really closes that file and the issue is gone.

Putting in another way, it appear that ESC works for closing any extra opened file expect the last one.
I don't know whether we should consider this as a bug but it is certainly not what the user wants to happen, so a probable solution would be to program the ESC to really close that last file and for the user side, a solution would be to avoid using ESC to close the last file.

So, I hope I contribute to clear up this little issue guys. Little but a bit annoying.

Thanks

**Reply by GitHubRulesOK** [reply]
you can try using the `EscToExit = false` command = true , and see it that helps your case?

**Reply by Peter-202122** [reply]
> However, it brings another issue for those using Sumatra with tabs turned off. So, whenever I opened one file it also opens a second windows and re-opens the last file previously opened in Sumatra, and of course that is not what the user wants. > 

That's a similar problem like the one I had too. ;-) (see https://github.com/sumatrapdfreader/sumatrapdf/issues/3953 )

You can try to solve it in the way mentioned above by GitHubRulesOK or in the way I did it (inspired also by GitHubRulesOK):

Change (in the SumatraPDF-settings.txt) the entry **"RestoreSession = true"** to **"RestoreSession = false"**

**Reply by morningscifi** [reply]
> Change (in the SumatraPDF-settings.txt) the entry **"RestoreSession = true"** to **"RestoreSession = false"**

Thanks very much, this was the missing piece.

This combination of settings got the behavior that I was hoping for:

1. Settings -> Options -> Tick the box for "remember opened files"
2. Settings -> Advanced Options -> Set `RestoreSession = false` in a text editor

No more ghost files popping up, and it remembers my page numbers. Perfect.

**Comment by marconedev** [comment]
The setting RememberStatePerDocument = true only works if RestoreSession = true is also enabled.
However, when both RememberStatePerDocument = true and RestoreSession = true are enabled, SumatraPDF exhibit problematic behavior.

Video Showing the BUG: https://www.youtube.com/watch?v=3MqKH9C96Cs

**Comment by marconedev** [comment]
![bug no sumatra](https://github.com/user-attachments/assets/54378c2a-e368-4dbb-adfe-f1e2cc8b0056)
